pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java

```diff
@@ -292,6 +250,69 @@ public void addFailed(ManagedLedgerException exception, Object ctx) {
         return completableFuture;
     }
 
+    private void handleMetadataEntry(PositionImpl logPosition, PendingAckMetadataEntry pendingAckMetadataEntry) {
+        // store the persistent position in to memory
+        // store the max position of this entry retain
+        if (pendingAckMetadataEntry.getPendingAckOp() != PendingAckOp.ABORT
+                && pendingAckMetadataEntry.getPendingAckOp() != PendingAckOp.COMMIT) {
+            Optional<PendingAckMetadata> optional = pendingAckMetadataEntry.getPendingAckMetadatasList()
+                    .stream().max((o1, o2) -> ComparisonChain.start().compare(o1.getLedgerId(),
+                            o2.getLedgerId()).compare(o1.getEntryId(), o2.getEntryId()).result());
+
+            optional.ifPresent(pendingAckMetadata -> {
+                PositionImpl nowPosition = PositionImpl.get(pendingAckMetadata.getLedgerId(),
+                        pendingAckMetadata.getEntryId());
+
+                if (nowPosition.compareTo(maxAckPosition) > 0) {
+                    maxAckPosition = nowPosition;
+                }
+                if (currentIndexLag.get() >= maxIndexLag) {
+                    pendingAckLogIndex.compute(maxAckPosition,
+                            (thisPosition, otherPosition) -> logPosition);
+                    maxIndexLag = logIndexBackoff.next(pendingAckLogIndex.size());
+                    currentIndexLag.set(0);
+                }
+            });
+        }
+    }
+
+    private void clearUselessLogData() {
+        if (!pendingAckLogIndex.isEmpty()) {
+            PositionImpl deletePosition = null;
+            while (!pendingAckLogIndex.isEmpty()
+                    && pendingAckLogIndex.firstKey() != null
+                    && subManagedCursor.getPersistentMarkDeletedPosition() != null
+                    && pendingAckLogIndex.firstEntry().getKey()
+                    .compareTo((PositionImpl) subManagedCursor.getPersistentMarkDeletedPosition()) <= 0) {
+                deletePosition = pendingAckLogIndex.firstEntry().getValue();
+                pendingAckLogIndex.remove(pendingAckLogIndex.firstKey());
+            }
+
+            if (deletePosition != null) {
+                maxIndexLag = logIndexBackoff.next(pendingAckLogIndex.size());
```

Comment:
Do we need to update the `maxIndexLag` here? We can just rely on `handleMetadataEntry()` to update the `maxIndexLag`?

Making only one place to update the `maxIndexLag` will keep the implementation simple.

Reply:
Consider such a case:
Due to a transaction do not end for a long time, the mark delete position do not move and the pendingAckLogIndex have many indexes. This means maxIndexLag is very big.
When the transaction end and pendingAckLogIndex was cleared, we should make maxIndexLag be minLag.
If not, the current lag still need to reach a big lag, which make the data impossible to clear for a long time.

Reply:
Ok, so do we need to consider if the updated `maxIndexLag` will < `currentIndexLag`?

Reply:
Of course, so we make the judge` currentIndexLag.get() >= maxIndexLag`